stdlib/http/src/main/java/org/ballerinalang/net/http/websocket/client/InitEndpoint.java

```diff
@@ -59,140 +57,62 @@ public static void initEndpoint(ObjectValue webSocketClient) {
-        } else {
-            WebSocketUtil.establishWebSocketConnection(webSocketClient, wsService);
-        }
-
-    }
-
-    private static void waitForHandshake(CountDownLatch countDownLatch) {
-        try {
-            // Wait to call countDown()
-            countDownLatch.await();
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new WebSocketException(WebSocketConstants.ERROR_MESSAGE + e.getMessage());
-        }
-    }
-
-    private static void populateClientConnectorConfig(MapValue<String, Object> clientEndpointConfig,
-                                                      WebSocketClientConnectorConfig clientConnectorConfig,
-                                                      String scheme) {
-        clientConnectorConfig.setAutoRead(false); // Frames are read sequentially in ballerina.
-        clientConnectorConfig.setSubProtocols(WebSocketUtil.findNegotiableSubProtocols(clientEndpointConfig));
-        @SuppressWarnings(WebSocketConstants.UNCHECKED)
-        MapValue<String, Object> headerValues = (MapValue<String, Object>) clientEndpointConfig.getMapValue(
-                WebSocketConstants.CLIENT_CUSTOM_HEADERS_CONFIG);
-        if (headerValues != null) {
-            clientConnectorConfig.addHeaders(getCustomHeaders(headerValues));
-        }
-
-        long idleTimeoutInSeconds = WebSocketUtil.findTimeoutInSeconds(clientEndpointConfig,
-                WebSocketConstants.ANNOTATION_ATTR_IDLE_TIMEOUT, 0);
-        if (idleTimeoutInSeconds > 0) {
-            clientConnectorConfig.setIdleTimeoutInMillis((int) (idleTimeoutInSeconds * 1000));
-        }
-
-        clientConnectorConfig.setMaxFrameSize(WebSocketUtil.findMaxFrameSize(clientEndpointConfig));
-
-        MapValue secureSocket = clientEndpointConfig.getMapValue(HttpConstants.ENDPOINT_CONFIG_SECURE_SOCKET);
-        if (secureSocket != null) {
-            HttpUtil.populateSSLConfiguration(clientConnectorConfig, secureSocket);
-        } else if (scheme.equals(WebSocketConstants.WSS_SCHEME)) {
-            clientConnectorConfig.useJavaDefaults();
+        if (webSocketClient.getType().getName().equalsIgnoreCase(WebSocketConstants.WEBSOCKET_CLIENT)) {
+            if (WebSocketUtil.hasRetryContext(webSocketClient)) {
+                @SuppressWarnings(WebSocketConstants.UNCHECKED)
+                MapValue<String, Object> retryConfig = (MapValue<String, Object>) clientEndpointConfig.getMapValue(
+                        WebSocketConstants.RETRY_CONTEXT);
+                RetryContext retryConnectorConfig = new RetryContext();
+                populateRetryConnectorConfig(retryConfig, retryConnectorConfig);
+                webSocketClient.addNativeData(WebSocketConstants.RETRY_CONTEXT, retryConnectorConfig);
+                webSocketClient.addNativeData(WebSocketConstants.CLIENT_LISTENER, new RetryConnectorListener(
+                        new ClientConnectorListener()));
```

Comment:
I think now it is a good time to move these listener creations to their respective endpoint classes :)

Reply:
I have created the retryClient as per the offline discussion.